test/gurka/test_recost.cc

```diff
@@ -0,0 +1,627 @@
+                                         |     |
+                                         D--6--E--7--F)";
+  const gurka::ways ways = {
+      {"A1B23C", {{"highway", "residential"}}},
+      {"D6E7F", {{"highway", "residential"}}},
+      {"B4D", {{"highway", "trunk"}}},
+      {"C5E", {{"highway", "trunk"}}},
+  };
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 10);
+  std::string tile_dir = "test/data/gurka_recost";
+  auto map = gurka::buildtiles(layout, ways, {}, {}, tile_dir, build_config);
+  auto reader = std::make_shared<baldr::GraphReader>(map.config.get_child("mjolnir"));
+
+  // run all the permutations
+  std::string named_locations = "1A2B3C4D5E6F7";
+  std::vector<std::unordered_map<std::string, std::string>> options{
+      {},
+      {{"/date_time/type", "1"}, {"/date_time/value", "2020-06-09T14:12"}},
+      {{"/date_time/type", "2"}, {"/date_time/value", "2020-06-22T14:12"}},
+  };
+  for (const auto& option : options) {
+    for (size_t i = 0; i < named_locations.size(); ++i) {
+      for (size_t j = i + 1; j < named_locations.size(); ++j) {
+        // skip uninterestingly close routes
+        if (i == j) {
+          continue;
+        }
+
+        // get the api response out using the normal means
+        auto start = named_locations.substr(i, 1);
+        auto end = named_locations.substr(j, 1);
+        auto api = gurka::route(map, start, end, "auto", option, reader);
+        auto leg = api.trip().routes(0).legs(0);
+
+        // setup a callback for the recosting to get each edge
+        auto edge_itr = leg.node().begin();
+        sif::EdgeCallback edge_cb = [&edge_itr]() -> baldr::GraphId {
+          auto edge_id =
+              edge_itr->has_edge() ? baldr::GraphId(edge_itr->edge().id()) : baldr::GraphId{};
+          ++edge_itr;
+          return edge_id;
+        };
+
+        // setup a callback for the recosting to tell us about the new label each made
+        auto elapsed_itr = leg.node().begin();
+        double length = 0;
+        uint32_t pred = baldr::kInvalidLabel;
+        sif::LabelCallback label_cb = [&elapsed_itr, &length,
+                                       &pred](const sif::EdgeLabel& label) -> void {
+          length += elapsed_itr->edge().length() * 1000.0;
+          EXPECT_EQ(elapsed_itr->edge().id(), label.edgeid());
+          EXPECT_EQ(pred++, label.predecessor());
+          EXPECT_EQ(static_cast<uint8_t>(elapsed_itr->edge().travel_mode()),
+                    static_cast<uint8_t>(label.mode()));
+          EXPECT_NEAR(length, label.path_distance(), 2);
+          EXPECT_NEAR(elapsed_itr->cost().transition_cost().seconds(), label.transition_cost().secs,
+                      .1);
+          // TODO: test restrictions
+          // we need to move to the next node which has the elapsed time at the end of the edge
+          ++elapsed_itr;
+          EXPECT_NEAR(elapsed_itr->cost().elapsed_cost().seconds(), label.cost().secs, .1);
```

Comment:
here we test only that the transition.secs and elapsed.secs are the same between route and recost, we should also test that the costs are the same